menhir-recover/recovery_new.ml

```diff
@@ -0,0 +1,196 @@
+
+      [cases] is a mapping that associates to each possible state found at
+      stack.[-prefix]
+      (or None if the stack is empty) a list of reductions to execute.
+
+      The actual list of actions to reduce an item [(state, prod, pos)] is
+      given by
+          [Synthesizer.solution (Trail (state, prod, pos))]
+    *)
+
+  val recover : G.lr1 -> recovery
+  val report : Format.formatter -> unit
+end
+ 
+module Recover (G : GRAMMAR) (S : SYNTHESIZER with module G := G)
+  : RECOVERY with module G := G =
+struct
+  open G
+  open Utils
+
+  type item = lr1 * production * int
+
+  type recovery = {
+    prefix: int;
+    cases: (G.lr1 option * item list) list;
+  }
+
+  type trace = { cost : Cost.t; items : item list }
+
+  module Trace = struct
+    type t = trace
+    let min tr1 tr2 =
+      Cost.arg_min (fun t -> t.cost) tr1 tr2
+
+    let cat tr1 tr2 =
+      { cost = Cost.add tr1.cost tr2.cost; items = tr1.items @ tr2.items }
+  end
+
+  let depth = Lr1.tabulate (fun st ->
+    List.fold_left (fun pos (_, pos') -> max pos pos') 0 @@ Lr0.items (Lr1.lr0 st))
+
+  module State = struct
+    module S = Set.Make (
+      struct
+        let compare = Pervasives.compare
+        type t = Cost.t * Nonterminal.t
+      end)
+
+    type t = S.t ref * Trace.t option array
+
+    let empty () = (
+        ref S.empty, 
+        Array.make Nonterminal.count @@ Some {cost = Cost.infinite; items = []}
+      )
+
+    let is_empty (set, _) = S.is_empty !set
+
+    let get_min (set, arr) =
+      let (cost, nt) as elem = S.min_elt !set in
+      let i = Nonterminal.to_int nt in
+      let res = nt, Option.get arr.(i) in
+      arr.(i) <- None;
+      set := S.remove elem !set;
+      res
+
+    let add_elem (set, arr) (nt, tr) =
+      let i = Nonterminal.to_int nt in
+      match arr.(i) with
+        | Some tr' -> 
+          if tr.cost < tr'.cost then
+            (set := S.add (tr.cost, nt) @@ S.remove (tr'.cost, nt) !set;
+            arr.(i) <- Some tr)
+        | _ -> ()
+  end
+
+  let synthesize = 
+    Lr1.tabulate (fun st ->
+      List.fold_left (fun acc (prod, pos) ->
+        let item = st, prod, pos in
+        let cost, _ = S.solve (S.Tail(st, prod, pos)) in
+        (Production.lhs prod, {cost = cost; items = [item]}) :: acc
+      ) [] (Lr0.items (Lr1.lr0 st))
+    )
+
+  let step child parent =
+    let finish = List.filter_map (fun (prod, pos) ->
+        if Array.length (Production.rhs prod) <= pos then
+          None
+        else 
+          match (Production.rhs prod).(pos) with
+            | (N nt, _, _) -> 
+              let next = List.assoc (N nt) (Lr1.transitions parent) in
+              let cost, _ = S.solve (S.Tail(next, prod, pos + 1)) in
+              Some (nt, {cost = cost; items = [next, prod, pos + 1]})
+            | _ -> None
+      ) @@ Lr0.items (Lr1.lr0 parent) in
+    
+    let state = State.empty () in
+
+    List.iter (State.add_elem state) @@ synthesize child;
+
+    let rec aux acc =
```

Comment:
please rename it or/and add comments

Reply:
I thought that name `aux` is just a common name for main sub function in OCaml

Reply:
> I thought that name aux is just a common name for main sub function in OCaml

Yeah, like `a`, `b`, `foo` and etc. But I think we should avoid it.
Since there aren't any comments at least names should be descriptive.

Reply:
Fixed